react-components/src/apps/admin/AdminAppDetails.js

```diff
@@ -0,0 +1,314 @@
+                multiline
+                rows={15}
+                component={FormTextField}
+            />
+        </Form>
+    );
+}
+
+function validateAppName(restrictedStartingChars, restrictedChars, value) {
+    if (!value) {
+        return "Empty value";
+    }
+
+    let startingCharsRegex = new RegExp("^[" + restrictedStartingChars + "]");
+    let invalid = value.match(startingCharsRegex);
+    if (invalid) {
+        return (
+            "App name cannot begin with `" +
+            restrictedStartingChars +
+            "`. Invalid char: `" +
+            invalid +
+            "`"
+        );
+    }
+
+    // Escape each non-alphanumeric char since some are used as special chars in regex
+    let escapedRestrictedChars = restrictedChars.replace(/\W/g, "\\$&");
+    let restrictedCharsRegex = new RegExp(
+        "[" + escapedRestrictedChars + "]",
+        "g"
+    );
+    invalid = value.match(restrictedCharsRegex);
+    if (invalid) {
+        return (
+            "App name cannot contain `" +
+            restrictedChars +
+            "`. Invalid chars: `" +
+            invalid.join("") +
+            "`"
+        );
+    }
+}
+
+const handleSubmit = (values, { props, setSubmitting }) => {
+    const { presenter } = props;
+
+    let promises = [];
+    if (props.app !== values.app) {
+        let saveApp = new Promise((resolve, reject) => {
+            presenter.onSaveAppSelected(values.app, resolve, reject);
+        });
+        promises.push(saveApp);
+    }
+
+    if (props.app.beta !== values.app.beta) {
+        let betaUpdate = new Promise((resolve, reject) => {
+            presenter.updateBetaStatus(values.app, resolve, reject);
+        });
+        promises.push(betaUpdate);
+    }
+
+    if (!props.documentation && values.documentation) {
+        let addAppDoc = new Promise((resolve, reject) => {
+            presenter.addAppDocumentation(
+                values.app.system_id,
+                values.app.id,
+                values.documentation,
+                resolve,
+                reject
+            );
+        });
+        promises.push(addAppDoc);
+    } else if (props.documentation !== values.documentation) {
+        let updateAppDoc = new Promise((resolve, reject) => {
+            presenter.updateAppDocumentation(
+                values.app.system_id,
+                values.app.id,
+                values.documentation,
+                resolve,
+                reject
+            );
+        });
+        promises.push(updateAppDoc);
+    }
+
+    Promise.all(promises)
+        .then(() => {
+            setSubmitting(false);
+            props.presenter.closeAppDetailsDlg();
+        })
+        .catch(() => {
+            setSubmitting(false);
+        });
+};
+
+function mapPropsToValues(props) {
+    const { app, documentation } = props;
+    return {
+        app: app,
+        documentation: documentation,
+    };
```

Comment:
Similar to destructuring, the key doesn't have to be specified here if it's the same as the variable name:
```javascript
     return {
        app,
        documentation,
    };
```
And to get really terse, we can do away with this function definition all together and use an `=>` function below:
```javascript
 export default withFormik({
    enableReinitialize: true,
    mapPropsToValues: ({ app, documentation }) => ({ app, documentation }),
    handleSubmit,
})(withI18N(AdminAppDetailsDialog, messages));
```

Reply:
Ah, terse is nicer!
I end up needing to modify this a bit in #438, so I'll save the update for that PR